driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -271,11 +267,36 @@ public void onResult(final BsonDocument result, final Throwable t) {
         public void onResult(final QueryResult<T> result, final Throwable t) {
             if (t != null) {
                 connection.release();
+                connectionSource.release();
                 close();
                 callback.onResult(null, t);
             } else {
                 handleGetMoreQueryResult(connection, callback, result);
             }
         }
     }
+
+    private ServerCursor getCursorForNext() {
+        ServerCursor localCursor;
+        synchronized (lock) {
+            localCursor = cursor.get();
+            if (localCursor == null) {
```

Comment:
Is it necessary to close the cursor at this point?  Can this be simplified to:

     if (localCursor != null) {
           connectionSource.retain();
     }

I think so since the next method already calls close() if this method returns a null cursor.